CSSFontFaceSet.cpp: Fix unsafe references to ref-counted argument

vitorroriz opened the pull request:
#### 410d814e111dacbd485c5b6fa2d10dd6c53974b2
<pre>
CSSFontFaceSet.cpp: Fix unsafe references to ref-counted argument
<a href="https://rdar.apple.com/148775146">rdar://148775146</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=291232">https://bugs.webkit.org/show_bug.cgi?id=291232</a>

Reviewed by Ryosuke Niwa.

* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/css/CSSFontFaceSet.cpp:
(WebCore::CSSFontFaceSet::updateStyleIfNeeded):
(WebCore::CSSFontFaceSet::ensureLocalFontFacesForFamilyRegistered):
(WebCore::CSSFontFaceSet::remove):
(WebCore::CSSFontFaceSet::matchingFacesExcludingPreinstalledFonts):
* Source/WebCore/css/CSSFontFaceSet.h:

Canonical link: <a href="https://commits.webkit.org/293613@main">https://commits.webkit.org/293613@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43036a60fc6b7606701e3a40bff624081457e02b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99380 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19029 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9281 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104511 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49981 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101421 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19318 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27463 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75635 "Found 1 new test failure: imported/w3c/web-platform-tests/service-workers/service-worker/redirected-response.https.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32735 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102387 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14687 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89725 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55994 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14481 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7708 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49341 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84407 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7795 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106868 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26494 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/19325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84596 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26856 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85929 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84109 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21341 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28777 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6469 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20237 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26434 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31635 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26254 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29567 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27821 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/css/CSSFontFaceSet.h:
```diff
@@ -143,7 +143,7 @@ class CSSFontFaceSet final : public RefCounted<CSSFontFaceSet>, public CSSFontFa
     WeakHashSet<FontModifiedObserver> m_fontModifiedObservers;
     WeakHashSet<FontEventClient> m_fontEventClients;
-    WeakPtr<CSSFontSelector> m_owningFontSelector;
+    const WeakPtr<CSSFontSelector> m_owningFontSelector;
```
A constant weak pointer is kind of a funny idea. Even if you don’t reassign it, it can reassign itself at any time. What’s the goal of marking it constant?

vitorroriz replied:
You are right, we need a protected accessor

geoffreygaren commented on Source/WebCore/css/CSSFontFaceSet.cpp:
```diff
@@ -268,10 +268,11 @@ void CSSFontFaceSet::remove(const CSSFontFace& face)
 
     for (size_t i = 0; i < m_faces.size(); ++i) {
-        if (m_faces[i].ptr() == &face) {
+        Ref protectedFace = m_faces[i];
```
Can we avoid the recount in the case where the pointers do not compare equal?

rniwa commented on Source/WebCore/css/CSSFontFaceSet.cpp:
```diff
@@ -117,19 +117,21 @@ void CSSFontFaceSet::ensureLocalFontFacesForFamilyRegistered(const AtomString& f
         return;
 
-    if (!m_owningFontSelector->scriptExecutionContext())
+    Ref protectedOwningFontSelector { *m_owningFontSelector };
```
We typically don't call these protectedFoo anymore. Also, we usually use = instead.

rniwa commented on Source/WebCore/css/CSSFontFaceSet.cpp:
```diff
@@ -117,19 +117,21 @@ void CSSFontFaceSet::ensureLocalFontFacesForFamilyRegistered(const AtomString& f
+    if (!protectedOwningFontSelector->scriptExecutionContext())
         return;
-    auto allowUserInstalledFonts = m_owningFontSelector->scriptExecutionContext()->settingsValues().shouldAllowUserInstalledFonts ? AllowUserInstalledFonts::Yes : AllowUserInstalledFonts::No;
+    Ref scriptExecutionContext { *protectedOwningFontSelector->scriptExecutionContext() };
```
We should probably add `CSSFontSelector::protectedScriptExecutionContext()` and use it here.

rniwa commented on Source/WebCore/css/CSSFontFaceSet.cpp:
```diff
@@ -418,7 +420,8 @@ ExceptionOr<Vector<std::reference_wrapper<CSSFontFace>>> CSSFontFaceSet::matchin
                 ASSERT(m_owningFontSelector && m_owningFontSelector->scriptExecutionContext());
-                familyAtom = AtomString { m_owningFontSelector->scriptExecutionContext()->settingsValues().fontGenericFamilies.standardFontFamily() };
+                Ref protectedOwningFontSelector { *m_owningFontSelector };
+                familyAtom = AtomString { Ref { *protectedOwningFontSelector->scriptExecutionContext() }->settingsValues().fontGenericFamilies.standardFontFamily() };
```
and here.

vitorroriz committed:
CSSFontFaceSet.cpp: Fix unsafe references to ref-counted argument
rdar://148775146
https://bugs.webkit.org/show_bug.cgi?id=291232

Reviewed by Ryosuke Niwa.

* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/css/CSSFontFaceSet.cpp:
(WebCore::CSSFontFaceSet::updateStyleIfNeeded):
(WebCore::CSSFontFaceSet::ensureLocalFontFacesForFamilyRegistered):
(WebCore::CSSFontFaceSet::remove):
(WebCore::CSSFontFaceSet::matchingFacesExcludingPreinstalledFonts):
* Source/WebCore/css/CSSFontFaceSet.h:

Canonical link: https://commits.webkit.org/293613@main
